app/models/comment.rb

```diff
@@ -30,7 +30,7 @@ class Comment < ApplicationRecord
   before_validation :evaluate_markdown, if: -> { body_markdown }
   before_save :set_markdown_character_count, if: :body_markdown
   before_save :synchronous_spam_score_check
-  before_create :adjust_comment_parent_based_on_depth
+  before_create :check_if_commentable_present, :adjust_comment_parent_based_on_depth
```

Comment:
Can this be reworked as a validation on create instead of an exception raised?

Reply:
There is already a [validation rule](https://github.com/forem/forem/blob/main/app/models/comment.rb#L49) that validates the commentable_id but it is only applied when there is a commentable_type. Did you mean that we need to add a validation constraint ([little similar to this one](https://github.com/forem/forem/blob/main/app/models/article.rb#L89))which adds an error if the commentable is blank ?

Reply:
@rhymes I added a similar validation rule as you've mentioned but it did not seem to work either and the same error message "undefined method `id' for nil:NilClass`" is shown in the error modal.  

![validation](https://raw.githubusercontent.com/saroj990/img-gallery/main/Screenshot%202021-05-13%20at%201.12.38%20AM.png)

But when I create a new comment record from the rails console, I can see the error is shown there. I guess validation errors are somehow not caught by the standard error handler.

![irb](https://raw.githubusercontent.com/saroj990/img-gallery/main/Screenshot%202021-05-13%20at%201.25.26%20AM.png)
 
This [method](https://github.com/forem/forem/blob/main/app/models/notification.rb#L61) is the real culprit here. It tries to refer to the commentable relation there but it seems to be nil by that time.

Reply:
I only added `validates :commentable, on: :create, presence: true` to the Comment model (without commenting the other validation which should remain):

```ruby
> cm = Comment.create(commentable: nil, user: user, body_markdown: "another ocmmcn")
> cm.save!
ActiveRecord::RecordInvalid: Validation failed: Commentable can't be blank
> cm.errors
 @errors=[#<ActiveModel::Error attribute=commentable, type=blank, options={:on=>:create}>]>
```

it seems correct to me! 

If I start the app, open an article's page I see the error, but I think the error comes from this line here:

https://github.com/forem/forem/blob/3c924193117ee0c8a1c3f66322f0335aa60ea053/app/controllers/comments_controller.rb#L86

I changed that line to: `commentable_id: @comment.commentable_id` and I think it worked:

![Screenshot 2021-05-13 at 9 33 29 AM](https://user-images.githubusercontent.com/146201/118093937-668bb980-b3ce-11eb-8c29-897e34194f19.png)

The error message should probably remain generic because comments can be attached to other things like podcast episodes, maybe something like derived from the value of `commentable_type` ?